src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -351,35 +375,56 @@ trait Contexts { self: Analyzer =>
         val pruned = prune(List(result.tree), implicitDictionary.map(_._2), Nil)
         if(pruned.isEmpty) result
         else {
-          val (vsyms, vdefs) = pruned.map { case (vsym, rhs) =>
-            (vsym, ValDef(Modifiers(FINAL | SYNTHETIC), vsym.name.toTermName, TypeTree(rhs.tpe), rhs.changeOwner(owner -> vsym)))
-          }.unzip
-
-          val mname = newTermName(typer.fresh.newName("LazyDefns$"))
-          val mdef =
-            ModuleDef(Modifiers(SYNTHETIC), mname,
-              gen.mkTemplate(gen.mkParents(NoMods, Nil), noSelfType, NoMods, ListOfNil, vdefs)
+          val vdefs = pruned.map { case (vsym, rhs) =>
+            changeNonLocalOwners(rhs, vsym)
+            ValDef(Modifiers(FINAL | SYNTHETIC), vsym.name.toTermName, TypeTree(rhs.tpe), rhs)
+          }
+
+          val cname = newTypeName(typer.fresh.newName("LazyDefns$"))
+          val cdef0 =
+            ClassDef(Modifiers(FINAL | SYNTHETIC), cname, Nil,
```

Comment:
If possible, it would be preferable to use an approach closer to the one in synthesizePartialFunction (https://github.com/scala/scala/blob/036143bad49e07bf131386bf2df39ec889e24dc9/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L2805-L2814), and call the ClassDef factory method at https://github.com/scala/scala/blob/80f165a631c675eeb593031f8cff541c7e8ca15e/src/compiler/scala/tools/nsc/ast/Trees.scala#L61